src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java

```diff
@@ -913,6 +914,13 @@ void move(UnixPath source, UnixPath target, CopyOption... options)
             if (!flags.replaceExisting) {
                 throw new FileAlreadyExistsException(
                     target.getPathForExceptionMessage());
+            } else if (!sourceAttrs.isSymbolicLink() || flags.followLinks) {
+                // ensure source can be moved
+                try {
+                    access(source, R_OK|W_OK);
```

Comment:
I wonder this should be changed to R_OK and done after the rename attempt has failed. I think that would make it consistent with the copy case (as move falls back to a copy when rename fails).

Reply:
But the existing target would be deleted after line 928 before rename() is called.

Reply:
Okay, you're right but I'm not 100% sure that the mode is right, I assumed it would be R_OK.

Reply:
> I will investigate the mode.

Okay. The change to copy looks okay, as does the change to the helper for cross-provider copy/move, it's just the move in the Unix provider where I can't say for sure without doing some experiments.

Reply:
The `W_OK` is not needed here as you suggested. With only `R_OK` being set, the proposed patch does not delete the existing target, but the mainline does.